src/share/cxx/CamForcing.cpp

```diff
@@ -0,0 +1,190 @@
+    const ExecViewManaged<Scalar * [QSIZE_D][NP][NP][NUM_LEV]> &Q) {
+
+  const int num_e = ps_v.extent_int(0);
+
+  const auto policy = Homme::get_default_team_policy<ExecSpace>(num_e);
+
+  Kokkos::parallel_for(
+      "tracer qdp forcing",
+      Kokkos::RangePolicy<ExecSpace>(0, num_e * num_q * NP * NP * NUM_LEV),
+      KOKKOS_LAMBDA(const int & idx) {
+        const int ie = (((idx / NUM_LEV) / NP) / NP) / num_q;
+        const int iq = (((idx / NUM_LEV) / NP) / NP) % num_q;
+        const int igp = ((idx / NUM_LEV) / NP) % NP;
+        const int jgp = (idx / NUM_LEV) % NP;
+        const int k = idx % NUM_LEV;
+        Scalar v1 = f_q(ie, iq, igp, jgp, k);
+        Scalar &qdp_s = qdp(ie, tl.np1_qdp, iq, igp, jgp, k);
+        VECTOR_SIMD_LOOP
+        for (int vlev = 0; vlev < VECTOR_SIZE; ++vlev) {
+          if (qdp_s[vlev] + v1[vlev] * dt < 0.0 && v1[vlev] < 0.0) {
+            if (qdp_s[vlev] < 0.0) {
+              v1[vlev] = 0.0;
+            } else {
+              v1[vlev] = -qdp_s[vlev] / dt;
+            }
+          }
+          qdp_s[vlev] += v1[vlev];
+        }
+      });
+
+  if (moisture == MoistDry::MOIST) {
+    // Remove the m_fq_ps_v buffer since it's not actually needed.
+    // Instead apply the forcing to m_ps_v directly
+    // Bonus - one less parallel reduce in dry cases!
+
+    // This conserves the dry mass in the process of forcing tracer 0
+    Kokkos::parallel_for("tracer FQ_PS forcing", policy,
+                         KOKKOS_LAMBDA(const TeamMember & team) {
+      KernelVariables kv(team);
+      const int &ie = kv.team_idx;
+
+      Kokkos::parallel_for(Kokkos::TeamThreadRange(kv.team, NP * NP),
+                           [&](const int pt_idx) {
+        const int igp = pt_idx / NP;
+        const int jgp = pt_idx % NP;
+
+        Real ps_v_forcing = 0.0;
+
+        Dispatch<ExecSpace>::parallel_reduce(
+            kv.team, Kokkos::ThreadVectorRange(kv.team, NUM_PHYSICAL_LEV),
+            [&](const int &k, Real &accumulator) {
+              const int &&ilev = k / VECTOR_SIZE;
+              const int &&vlev = k % VECTOR_SIZE;
+              double v1 = f_q(ie, 0, igp, jgp, ilev)[vlev];
+              const double &qdp_s =
+                  qdp(ie, tl.np1_qdp, 0, igp, jgp, ilev)[vlev];
+              if (qdp_s + v1 * dt < 0.0 && v1 < 0.0) {
+                if (qdp_s < 0.0) {
+                  v1 = 0.0;
+                } else {
+                  v1 = -qdp_s / dt;
+                }
+              }
+              accumulator += v1;
+            },
+            ps_v_forcing);
+        ps_v(ie, tl.np1, igp, jgp) += ps_v_forcing;
+      });
+    });
+  }
+
+  Kokkos::parallel_for("tracer forcing ps_v",
+                       Kokkos::RangePolicy<ExecSpace>(0, num_e * num_q * NP *
+                                                             NP * NUM_LEV),
+                       KOKKOS_LAMBDA(const int & idx) {
+    const int ie = (((idx / NUM_LEV) / NP) / NP) / num_q;
```

Comment:
We could imagine at some point as a design issue making a bunch of `Homme::parallel_for`'s that hide this somewhat frequently repeated identical div-mod arithmetic and provide the functor with `(ie,iq,igp,jgp,k)` and variants.

Reply:
:+1: I've made issue #359 for you